pgmpy/models/DiscreteBayesianNetwork.py

```diff
@@ -1190,9 +1200,28 @@ def do(self, nodes, inplace=False):
         adj_model = DAG.do(model, nodes, inplace=inplace)
 
         if adj_model.cpds:
-            for node in nodes:
+            for index, node in enumerate(nodes):
                 cpd = adj_model.get_cpds(node=node)
-                cpd.marginalize(cpd.variables[1:], inplace=True)
+                state_names = cpd.state_names[node]
+                if states[index] not in state_names:
+                    raise ValueError(
+                        f"State {states[index]} not defined for {node} in the model."
+                    )
+                idx = state_names.index(states[index])
```

Comment:
TabularCPD object has a method `get_state_no(var, state_name)` that you can use here instead. It utilizes pre-computed state_name to number mapping to save compute.